### PR TITLE
Improve UI `update_clipping_system` comments

### PR DESCRIPTION
**Pull request by doup**
# Objective

- Improve `update_clipping_system` comments

## Solution

- Add extra comments
- Reorder `CalculatedClip` updating `match` so it's reads `(old, new)` vs previous `(new, old)`
- Remove `clone` on children iteration

### CODE REVIEW

File: crates/bevy_ui/src/update.rs [hunk 1]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
     entity: Entity,
     clip: Option<Rect>,
 ) {
-    let (node, global_transform, style, calculated_clip) = node_query.get_mut(entity).unwrap();
-    // Update this node's CalculatedClip component
-    match (clip, calculated_clip) {
+    let (node, global_transform, style, old_clip) = node_query.get_mut(entity).unwrap();
+
+    // Update current node's CalculatedClip component
+    match (old_clip, clip) {
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:43]: `(old_clip, clip)` is an improvement but I think the semantics would be even clearer if you renamed `clip` as well so it's 

```suggestion
    match (old_clip, new_clip) {
```

File: crates/bevy_ui/src/update.rs [hunk 2]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
...
+        (None, Some(clip)) => {
             commands.entity(entity).insert(CalculatedClip { clip });
         }
-        (Some(clip), Some(mut old_clip)) => {
+        (Some(mut old_clip), Some(clip)) => {
             if old_clip.clip != clip {
                 *old_clip = CalculatedClip { clip };
             }
         }
     }
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:56]: I was going to suggest to use `set_if_neq` but then I checked and we never actually query for changes in `CalculatedClip` so the neq gate is unnecessary: 

```suggestion
        (Some(mut old_clip), Some(clip)) =>  *old_clip = CalculatedClip { clip },
    }
```

File: crates/bevy_ui/src/update.rs [hunk 3]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
...
+    // Calculate new clip rectangle for children nodes
     let children_clip = match style.overflow {
+        // When `Visible`, children will be visible even when they are outside
+        // the current node's "area". In this case they inherit the current
+        // node's parent clip. If an ancestor is set as `Hidden`, that clip will
+        // be used; otherwise this will be `None`.
         Overflow::Visible => clip,
         Overflow::Hidden => {
+            // Calculate current node clip rectangle from: posisition + calculated_size
             let node_center = global_transform.translation().truncate();
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:67]: There is a helper function for this now:
```suggestion
            let node_clip = node.logical_rect(global_transform);
```

**Review by ickshonpe — CHANGES_REQUESTED** [review]
Looks good, just got a few suggestions for some other improvements.

File: crates/bevy_ui/src/update.rs [hunk 4]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
     entity: Entity,
     clip: Option<Rect>,
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:38]: This parameter could be renamed to make clear that the clipping `Rect` is inherited from the node's parent:

```suggestion
   maybe_inherited_clip: Option<Rect>,
```

File: crates/bevy_ui/src/update.rs [hunk 5]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
...
 ) {
-    let (node, global_transform, style, calculated_clip) = node_query.get_mut(entity).unwrap();
-    // Update this node's CalculatedClip component
-    match (clip, calculated_clip) {
+    let (node, global_transform, style, old_clip) = node_query.get_mut(entity).unwrap();
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:40]: I think `maybe_calculated_clip` is better than `old_clip`, makes it clearer it's an optional smart pointer to the component.

```suggestion
    let (node, global_transform, style, maybe_calculated_clip) = node_query.get_mut(entity).unwrap();
```

File: crates/bevy_ui/src/update.rs [hunk 6]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
...
+    // Update current node's CalculatedClip component
+    match (old_clip, clip) {
         (None, None) => {}
-        (None, Some(_)) => {
+        (Some(_), None) => {
             commands.entity(entity).remove::<CalculatedClip>();
         }
-        (Some(clip), None) => {
+        (None, Some(clip)) => {
             commands.entity(entity).insert(CalculatedClip { clip });
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:49]:
```suggestion
        (None, Some(inherited_clip)) => {
            commands.entity(entity).insert(CalculatedClip { clip: inherited_clip });
```

File: crates/bevy_ui/src/update.rs [hunk 7]
```diff
@@ -37,35 +37,46 @@ fn update_clipping(
...
         }
-        (Some(clip), Some(mut old_clip)) => {
+        (Some(mut old_clip), Some(clip)) => {
```
Comment (ickshonpe) [inline crates/bevy_ui/src/update.rs:51]:
```suggestion
        (Some(mut calculated_clip), Some(inherited_clip)) => {
```

**Review by ickshonpe — CHANGES_REQUESTED** [review]

**Comment by ickshonpe** [comment]
#8095 that is very related needs some reviews, be grateful if you have a few minutes spare to look at it.

**Comment by doup** [comment]
Thanks for the comments! I was going to review them, but suddenly I'm getting a segfault out of nowhere. So frustrating. The good thing is that I can learn how this kind of stuff is debugged.

Anyway, sure I can check #8095. Btw, I want to review more UI PRs (mostly yours and nico's), but first I want to finish reading the UI related code. I did few reviews last couple of weeks but I realized I was lacking understanding. ^_^

**Comment by doup** [comment]
@ickshonpe I've applied most of the suggestions. I think I'm going to skip https://github.com/bevyengine/bevy/pull/8147#discussion_r1143108441 I prefer not to touch that part. In any case, I've asked about it in #8095, we can follow there.

**Review by mockersf — APPROVED** [review]

**Review by ickshonpe — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Pull request merged**